osbuild/__init__.py

```diff
@@ -374,19 +374,13 @@ def run(self, tree, build_tree, output_dir=None, interactive=False, check=True,
 
 
 class Pipeline:
-    def __init__(self, base=None):
-        self.base = base
-        self.build = None
+    def __init__(self, build=None):
+        self.build = build
         self.stages = []
         self.assembler = None
 
     def get_id(self):
-        return self.stages[-1].id if self.stages else self.base
-
-    def set_build(self, pipeline):
-        if self.stages:
-            raise ValueError("Must set build before stages.")
-        self.build = pipeline
+        return self.stages[-1].id if self.stages else None
```

Comment:
This method's signature has changed, I suggest to modify places where it is invoked.

according to my IDE, it is invoked here:
```
    def add_stage(self, name, options=None):
        build = self.build.get_id() if self.build else None
        stage = Stage(name, build, self.get_id(), options or {})
```
but the `__init__` method from the `Stage` class cannot handle this case:
```
class Stage:
    def __init__(self, name, base, build, options):
        m = hashlib.sha256()
        m.update(json.dumps(name, sort_keys=True).encode())
        m.update(json.dumps(build, sort_keys=True).encode())
        m.update(json.dumps(base, sort_keys=True).encode())
```
So I guess `base or {}` could go here.

Reply:
You are referring to `Pipeline.get_id()`? I don't see how the signature changed? But I think you are right regardless that callers use it wrong.

Reply:
Actually, except for a confusion of the order of he arguments, I don't see the problem. What o you have in mind?

Reply:
My bad, the `self.base` could have been `None` as well so there should be no difference.